### PR TITLE
Add support for custom providers

### PR DESCRIPTION
**Pull request by Xirado**
Adds support for setting custom identifiers

### CODE REVIEW

**Comment by Xirado** [comment]
Oh this only supports Youtube atm

File: src/main/java/com/github/topisenpai/plugin/spotify/SpotifyConfig.java [hunk 1]
```diff
@@ -8,9 +8,16 @@
 @Component
 public class SpotifyConfig{
 
+	public static final String ISRC_PATTERN = "%ISRC%";
```
Comment (TopiSenpai) [inline src/main/java/com/github/topisenpai/plugin/spotify/SpotifyConfig.java:11]: this should not be in the config

File: src/main/java/com/github/topisenpai/plugin/spotify/SpotifyConfig.java [hunk 2]
```diff
@@ -8,9 +8,16 @@
...
+	public static final String QUERY_PATTERN = "%QUERY%";
```
Comment (TopiSenpai) [inline src/main/java/com/github/topisenpai/plugin/spotify/SpotifyConfig.java:12]: this too

File: src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java [hunk 3]
```diff
@@ -65,12 +72,30 @@ private String getQuery(){
...
 		AudioItem track = null;
-		if(this.isrc != null){
-			track = this.spotifySourceManager.getSearchSourceManager().loadItem(null, new AudioReference("ytsearch:\"" + this.isrc + "\"", null));
-		}
-		if(track == null){
-			track = this.spotifySourceManager.getSearchSourceManager().loadItem(null, new AudioReference(getQuery(), null));
+
+		for(int i = 0; i < config.providers.length; i++){
+			String identifier = config.providers[i];
+			if(identifier.startsWith(SpotifySourceManager.SEARCH_PREFIX)){
```
Comment (TopiSenpai) [inline src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java:80]: we should warn log here i think

File: src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java [hunk 4]
```diff
@@ -65,12 +72,30 @@ private String getQuery(){
...
+				continue;
+			}
+			if(this.isrc != null){
```
Comment (TopiSenpai) [inline src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java:83]: can't we do the contains check here? with the `isrc != null`?

File: src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java [hunk 5]
```diff
@@ -65,12 +72,30 @@ private String getQuery(){
...
+				identifier = identifier.replace(ISRC_PATTERN, this.isrc);
+			}else{
+				if(identifier.contains(ISRC_PATTERN)){
+					log.info("Spotify: Ignoring identifier \""+identifier+"\" because this track does not have an ISRC!");
```
Comment (TopiSenpai) [inline src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java:87]: this should not be logged under debug or else your console gets spammed with useless info

File: src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java [hunk 6]
```diff
@@ -65,12 +72,30 @@ private String getQuery(){
...
+					continue;
+				}
+			}
+			identifier = identifier.replace(QUERY_PATTERN, getQuery());
+			track = this.spotifySourceManager.getSearchSourceManager().loadItem(null, new AudioReference(identifier, null));
+			if (track != null){
+				log.info("Spotify: Loaded item from identifier \""+identifier+"\"");
```
Comment (TopiSenpai) [inline src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java:94]: why do you even want logs here

File: src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java [hunk 7]
```diff
@@ -65,12 +72,30 @@ private String getQuery(){
...
+				break;
+			}else{
+				log.info("Spotify: Identifier \""+identifier+"\" returned no results");
```
Comment (TopiSenpai) [inline src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java:97]: why an else after break?

File: src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java [hunk 8]
```diff
@@ -65,12 +72,30 @@ private String getQuery(){
...
+			if(this.isrc != null){
+				identifier = identifier.replace(ISRC_PATTERN, this.isrc);
+			}else{
+				if(identifier.contains(ISRC_PATTERN)){
+					log.info("Spotify: Ignoring identifier \""+identifier+"\" because this track does not have an ISRC!");
+					continue;
+				}
+			}
+			identifier = identifier.replace(QUERY_PATTERN, getQuery());
+			track = this.spotifySourceManager.getSearchSourceManager().loadItem(null, new AudioReference(identifier, null));
```
Comment (TopiSenpai) [inline src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java:92]: replace `getSearchSourceManager` with `getAudioPlayerManger` to support other sources like SoundCloud

File: src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java [hunk 9]
```diff
@@ -56,7 +63,7 @@ public String getArtworkURL(){
 	}
 
 	private String getQuery(){
```
Comment (TopiSenpai) [inline src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java:65]: this function should prob be renamed then

**Review by TopiSenpai — CHANGES_REQUESTED** [review]
there is an editorconfig which should auto format

File: src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java (hunk 8 above)
Comment (Xirado) [reply src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java:92]: The problem is that `AudioPlayerManager#loadItem` is async, but not sure if there's a way to do it blocking

Comment (TopiSenpai) [reply src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java:92]: guess why I was too lazy to implement custom providers

but without this the entire pr is pointless

**Comment by Xirado** [comment]
Okay, i'm using Reflection to get the List of source managers, not a good solution, but maybe i can PR a getter.

File: src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java [hunk 10]
```diff
@@ -55,8 +64,8 @@ public String getArtworkURL(){
 		return this.artworkURL;
 	}
 
-	private String getQuery(){
-		var query = "ytsearch:" + trackInfo.title;
+	private String buildSearchQuery(){
```
Comment (TopiSenpai) [inline src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java:67]:
```suggestion
	private String getTrackTitle(){
```

File: src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java [hunk 11]
```diff
@@ -65,12 +74,31 @@ private String getQuery(){
...
-			track = this.spotifySourceManager.getSearchSourceManager().loadItem(null, new AudioReference("ytsearch:\"" + this.isrc + "\"", null));
-		}
-		if(track == null){
-			track = this.spotifySourceManager.getSearchSourceManager().loadItem(null, new AudioReference(getQuery(), null));
+
+		for(int i = 0; i < config.providers.length; i++){
+			String identifier = config.providers[i];
+
+			if(identifier.startsWith(SpotifySourceManager.SEARCH_PREFIX)){
+				log.warn("Can not use spsearch as provider!");
```
Comment (TopiSenpai) [inline src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java:84]:
```suggestion
				log.warn("Can not use spotify search as provider");
```

File: src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java [hunk 12]
```diff
@@ -65,12 +74,31 @@ private String getQuery(){
...
 	public void process(LocalAudioTrackExecutor executor) throws Exception{
+		SpotifyConfig config = this.spotifySourceManager.getConfig();
 		AudioItem track = null;
-		if(this.isrc != null){
-			track = this.spotifySourceManager.getSearchSourceManager().loadItem(null, new AudioReference("ytsearch:\"" + this.isrc + "\"", null));
-		}
-		if(track == null){
-			track = this.spotifySourceManager.getSearchSourceManager().loadItem(null, new AudioReference(getQuery(), null));
+
+		for(int i = 0; i < config.providers.length; i++){
```
Comment (TopiSenpai) [inline src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java:80]:
```suggestion
		for(var provider : config.providers){
```

File: src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java [hunk 13]
```diff
@@ -65,12 +74,31 @@ private String getQuery(){
...
+			String identifier = config.providers[i];
```
Comment (TopiSenpai) [inline src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java:81]:
```suggestion
```

File: src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java [hunk 14]
```diff
@@ -65,12 +74,31 @@ private String getQuery(){
...
+			if(identifier.startsWith(SpotifySourceManager.SEARCH_PREFIX)){
+				log.warn("Can not use spsearch as provider!");
+				continue;
+			}
+
+			if(identifier.contains(ISRC_PATTERN)){
+				if(this.isrc != null){
+					identifier = identifier.replace(ISRC_PATTERN, this.isrc);
+				}else{
+					log.debug("Spotify: Ignoring identifier \"" + identifier + "\" because this track does not have an ISRC!");
```
Comment (TopiSenpai) [inline src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java:92]: why `Spotify:` out of sudden stay consistent with the other logs?

File: src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java [hunk 15]
```diff
@@ -88,4 +116,24 @@ public AudioSourceManager getSourceManager(){
 		return this.spotifySourceManager;
 	}
 
+	public AudioItem loadItem(String query){
```
Comment (TopiSenpai) [inline src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java:119]: I still don't like this reflection shit

File: src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java [hunk 16]
```diff
@@ -65,12 +74,31 @@ private String getQuery(){
...
+			if(identifier.contains(ISRC_PATTERN)){
+				if(this.isrc != null){
+					identifier = identifier.replace(ISRC_PATTERN, this.isrc);
+				}else{
+					log.debug("Spotify: Ignoring identifier \"" + identifier + "\" because this track does not have an ISRC!");
+					continue;
+				}
+			}
+
+			identifier = identifier.replace(QUERY_PATTERN, buildSearchQuery());
```
Comment (TopiSenpai) [inline src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java:97]:
```suggestion
			identifier = identifier.replace(QUERY_PATTERN, getTrackTitle());
```

File: src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java [hunk 17]
```diff
@@ -65,12 +74,31 @@ private String getQuery(){
...
+
+			if(identifier.startsWith(SpotifySourceManager.SEARCH_PREFIX)){
+				log.warn("Can not use spsearch as provider!");
+				continue;
+			}
+
+			if(identifier.contains(ISRC_PATTERN)){
+				if(this.isrc != null){
+					identifier = identifier.replace(ISRC_PATTERN, this.isrc);
+				}else{
```
Comment (TopiSenpai) [inline src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java:91]:
```suggestion
				}
				else{
```

File: src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java [hunk 18]
```diff
@@ -65,12 +74,31 @@ private String getQuery(){
 
 	@Override
 	public void process(LocalAudioTrackExecutor executor) throws Exception{
+		SpotifyConfig config = this.spotifySourceManager.getConfig();
```
Comment (TopiSenpai) [inline src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java:77]:
```suggestion
		var config = this.spotifySourceManager.getConfig();
```

**Review by TopiSenpai — CHANGES_REQUESTED** [review]
I won't merge before the reflection stuff is gone

**Review by TopiSenpai — CHANGES_REQUESTED** [review]
while this stays open for now until the reflection stuff is resolved, you can actually use it like this
```yaml
lavalink:
  plugins:
    - dependency: "com.github.topisenpai:spotify-plugin:PR4-SNAPSHOT"
      repository: "https://jitpack.io"
```

File: README.md [hunk 19]
```diff
@@ -26,6 +26,9 @@ plugins:
     clientId: "your client id"
     clientSecret: "your client secret"
     countryCode: "US" # the country code you want to use for filtering the artists top tracks. See https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2
+    providers: # Custom providers for track loading. This is the default
+        - "ytsearch:\"%ISRC%\"" # Will be ignored if track does not have an ISRC. See https://en.wikipedia.org/wiki/International_Standard_Recording_Code
+        - "ytsearch:%QUERY%" # Will be used if track has no ISRC
```
Comment (TopiSenpai) [inline README.md:31]:
```suggestion
        - "ytsearch:%QUERY%" # Will be used if track has no ISRC or no track could be found for the ISRC
       # - "scsearch:%QUERY%" you can add multiple other fallbacks sources here
```
the original comment was a bit unclear

File: README.md [hunk 20]
```diff
@@ -43,7 +47,7 @@ You can also use `spsearch:<query>`(remove the <>) to search for songs on Spotif
 
 ---
 
-In some cases a requested Spotify songs can't be found on YouTube then you will receive a normal `TrackStartEvent`
+In case a requested Spotify song can't be found on YouTube, you will receive a normal `TrackStartEvent`
```
Comment (TopiSenpai) [inline README.md:50]: this should not be YT specific anymore

File: src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java [hunk 21]
```diff
@@ -88,4 +116,37 @@ public AudioSourceManager getSourceManager(){
 		return this.spotifySourceManager;
 	}
 
+	public AudioItem loadItem(String query) throws Exception{
```
Comment (TopiSenpai) [inline src/main/java/com/github/topisenpai/plugin/spotify/SpotifyTrack.java:119]:
```suggestion
	public AudioItem loadItem(String query) throws ExecutionException, InterruptedException{
```

**Review by TopiSenpai — CHANGES_REQUESTED** [review]

**Review by TopiSenpai — APPROVED** [review]

**Pull request merged**